File: Cpp/sieugiaithua.cpp
#include <iostream>
#define ll long long
using namespace std;

ll count_zeros(ll x) {
    ll count = 0, i = 5;
    while (x / i > 0) {
        count += x / i;
        i *= 5;
    }
    return count;
}
ll solve(ll n) {
    ll lower_bound = 0, upper_bound = n * 5;
    while (lower_bound < upper_bound) {
        ll middle = (lower_bound + upper_bound) / 2;
        ll count = count_zeros(middle);
        if (count < n) {
            lower_bound = middle + 1;
        }
        else {
            upper_bound = middle;
        }
    }
    return lower_bound;
}
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    int n;
    cin >> n;
    cout << solve(n);
    return 0;
}

File: Python/sieugiaithua.py
def solve(n):
    def count_zeros(x):
        count = 0
        i = 5
        while x // i > 0:
            count += x // i
            i *= 5
        return count

    lower_bound = 0
    upper_bound = 5 * n

    while lower_bound < upper_bound:
        mid = (lower_bound + upper_bound) // 2
        count = count_zeros(mid)
        if count < n:
            lower_bound = mid + 1
        else:
            upper_bound = mid

    return lower_bound

for _ in range(1, 101):
    print(_, solve(_))